File: 0098_ValidateBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBSTFirst(TreeNode* root){
        if(root == NULL) return true;
        //TreeNode* root_copy = root;
        if (root->left != NULL) {
            if (root->right != NULL){
                if (root->left->val >= root->val || root->val >= root->right->val) return false;
            } else if (root->left->val >= root->val) return false;
            if (!isValidBSTSub(root->left)) return false;
        } else if (root->right != NULL) return (root->right->val > root->val);
        if (root->right != NULL) {
            if (root->left != NULL){
                if (root->left->val >= root->val || root->val >= root->right->val) return false;
            } else if (root->val >= root->right->val) return false;
            if(!isValidBSTSub(root->right)) return false;
        } else if (root->left != NULL) return (root->left->val < root->val);
        return true;
    }
    
    bool isValidBSTSub(TreeNode* root) {
        if(root == NULL) return true;
        //TreeNode* root_copy = root;
        if (root->left != NULL) {
            if (root->right != NULL){
                if (root->left->val >= root->val || root->val >= root->right->val) return false;
            } else if (root->left->val >= root->val) return false;
            if (!isValidBST(root->left)) return false;
        } else if (root->right != NULL) return (root->right->val > root->val);
        if (root->right != NULL) {
            if (root->left != NULL){
                if (root->left->val >= root->val || root->val >= root->right->val) return false;
            } else if (root->val >= root->right->val) return false;
            if(!isValidBST(root->right)) return false;
        } else if (root->left != NULL) return (root->left->val < root->val);
        return false;
    }
    
    bool isValidBST(TreeNode* root) {
        return isValidBSTFirst(root);
    }
};

File: 0098_ValidateBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxVal(self, root: TreeNode):
        max_val = root.val
        if root.left is not None:
            max_val = max(max_val, self.maxVal(root.left))
        if root.right is not None:
            max_val = max(max_val, self.maxVal(root.right))
        return max_val
    def minVal(self, root: TreeNode):
        min_val = root.val
        if root.left is not None:
            min_val = min(min_val, self.maxVal(root.left))
        if root.right is not None:
            min_val = min(min_val, self.maxVal(root.right))
        return min_val
    def isValidBST(self, root: TreeNode) -> bool:
        if root is None:
            return True
        if root.left is not None:
            if not self.maxVal(root.left) < root.val:
                return False
            if not self.isValidBST(root.left):
                return False
        if root.right is not None:
            if not root.val < self.minVal(root.right):
                return False
            if not self.isValidBST(root.right):
                return False
        return True
